ops/testing.py

```diff
@@ -394,9 +394,30 @@ def _create_meta(self, charm_metadata: Optional[YAMLStringOrFile],
         """Create a CharmMeta object.
 
         Handle the cases where a user doesn't supply explicit metadata snippets.
+        Will try to load metadata/config from charm_dir/charmcraft.yaml first, then
+        charm_dir/metadata.yaml and/or charm_dir/actions.yaml if charmcraft.yaml
+        not include metadata and/or actions.
         """
         filename = inspect.getfile(self._charm_cls)
         charm_dir = pathlib.Path(filename).parents[1]
+        charmcraft_yaml: Optional[str] = None
+        charmcraft_metadata: Optional[dict[str, Any]] = None
+
+        # Check charmcraft.yaml and load it if it exists
+        charmcraft_meta = charm_dir / "charmcraft.yaml"
+        if charmcraft_meta.exists():
+            self._charm_dir = charm_dir
+            charmcraft_yaml = charmcraft_meta.read_text()
+            charmcraft_metadata = yaml.safe_load(charmcraft_yaml)
+
+        # Add metadata from charmcraft.yaml if no metadata is provided
+        # Should fall back to metadata.yaml if both are not provided
+        if charm_metadata is None and charmcraft_metadata:
+            if any(
+                meta_key in charmcraft_metadata for meta_key in ("name", "summary", "description")
```

Comment:
A lot of this logic looks very similar to the logic in `main.py`. Ideally we can use `CharmMeta.from_files` or similar once that exists.

Reply:
We also need a couple of tests of this behaviour (for metadata and for config) in `test/test_testing.py`.

Reply:
Added tests.